### DISCUSSION TITLE
Improvements to user-based authentication for datasources

### DISCUSSION THREAD

**Discussion by kostrse** [grafana/grafana#39701] [Data Sources]
For use cases when access to datasource needs to be done in context of a current user Grafana offers option of **token pass-through** authentication.

The token pass-through authentication has a few problems:

1. Access tokens tend to expire and Grafana has no mechanism to refresh them. As result, dashboards opened for a long period of time eventually stop working and user has to do F5 refresh of the app to trigger re-login.
2. There's no way for a datasource to request a token with audience and/or scopes specific to this particular datasource, or provide it dynamically. Scopes have to be preconfigured in the Grafana server config.
3. Since there's no concept of on-demand token request, there's also no on-demand interactive flow (e.g. interactively give consent for a new datasource to give access to the data).

## Proposal

The proposal is to add support for datasource authentication with access tokens acquired on behalf of the current user, using scopes provided by the datasource (e.g. using on-demand access token request).

First, we could add a new authentication option to the `httpclient.Options` struct ([here](https://github.com/grafana/grafana-plugin-sdk-go/blob/91d830b3f5baaf85bbbc2ea2e2cb62820bfb73ed/backend/httpclient/options.go#L22)) without reference to any particular implementation:

```go
package httpclient

type Options struct {
    ...
    CurrentUserAuth: bool
    Scopes: []string
    ...
}
```

Second, in the `httpclientprovider` ([here](https://github.com/grafana/grafana/blob/9a178e7ba96c31be794e712bb098c7aa60309564/pkg/infra/httpclient/httpclientprovider/http_client_provider.go#L22)) we registrer middleware with implementation of the access token retrieval logic depending on configuration of the Grafana instance.

For example, if Grafana configured with OAuth user authentication, then a middleware would be registered which implements OAuth 2.0 on demand token requests, If Grafana configured with with [auth proxy](https://grafana.com/docs/grafana/latest/auth/auth-proxy/), then another middleware would be registered which supports token retrieval relevant for the auth proxy scenario.

If no user authentication method configured with access token support (e.g. local users), then datasource requests would fail with error indicating that authentication on behalf of the current user isn't supported.

By keeping the configuration agnostic to implementation would allow datasources to support the user-based authentication without relying on specifics authentication method. Same implementation of datasource would work regardless of how Grafana is configured.

Frontend would receive a configuration flag indicating whether the user-based auth is possible to enable/disable such option in the datasource configuration editor.

### Access token retrieval for OAuth-based authentication configurations

OAuth 2.0 providers support OBO flow (e.g. [OAuth 2.0 On-Behalf-Of flow in AAD](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow))

OAuth configurations already require `token_url` which is used for ID token retrieval during the login flow. The same endpoint can be used for the on-demand OBO requests triggered by datasources.

### Access token retrieval for non-OAuth configurations (e.g. auth proxy)

In scenarios when auth configuration doesn't have concept of access token requests, such as auth proxy, Grafana can support a generic token endpoint with a documented schema similar to a token endpoint from OAuth. Users would be able to implement such endpoint on their side and configure it with Grafana for their non-standard authentication scenarios if they need to:

```ini
########################## Generic Token Endpoint ##########################
[auth.generic_token_endpoint]
enabled = false
token_url = http://10.0.0.1/token
client_id =
client_secret =
...
```

### Token endpoint protocol

*TBD*

```http
POST http://10.0.0.1/token
Content-Type: application/x-www-form-urlencoded

client_id=...&
client_secret=...&
user_id=...&
scope=...
```

**Comment by kostrse** [comment]
@sunker, @joshhunt, @mckn

**Reply by mckn** [reply]
I think sounds great! I think @marefr should be included in this discussion.

**Comment by daniellee** [comment]
Hi @kostrse 

We have been having some internal discussions about this feature. It is something we want to work is on the long term backlog.

Here is the feedback from the team.

From @kminehart:

There's so many edge cases that aren't taken into consideration. Sergey is spot on in the discussion proposal with their bullets for why the oauth passthrough is not sufficient, but there are more reasons.

Here are some examples:

- The biggest one that often gets forgotten: **alerting**. What account is used to run alerts? Surely not the account of the person who created the alert, right? Cue service accounts (we have a design doc that we can share later). How are alerts even created and shared if different data could be presented to each user based on their auth context?

- **Other features**: I'm thinking specifically of recorded queries / grafana recording rules being implemented in Grafana Enterprise, which would have the exact same problem that alerting would, since it functions in a very similar way.

- **Caching**: This would break caching if the same query can return different data to different users, so I think data sources that implement this would have to have caching turned off.

- **Missing / bad UX**: How do we cleanly tell someone who's not authorized to "Please sign in with {provider}" on the dashboard without giving them an "Error: Unauthorized" in the panel?
Implementation / maintenance burden on data sources: Data sources that want to support this have to have even more boilerplate code to support this feature. The option may show up on some data sources that do not support it.

- **Lack of documentation / User Confusion**: While Microsoft may have their on "OBO" (on behalf-of) spec, other OIDC providers do not. Generally, if a user authenticated with OIDC Provider X in Grafana,  and they want to use that token in another service Y which also requires authentication via OIDC provider X, then we have a scenario called [cross-JWT confusion](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-jwt-bcp#section-2.8). I've seen so many requests now, lots in GitHub issues and in StackOverflow in the grafana tag, where users want to use OAuth token passthrough for this. This doesn't necessarily apply to Microsoft's point, since it looks like their OBO specification is intended to be used in this scenario (see this chart here: https://docs.microsoft.com/en-us/azure/active-directory/develop/media/v2-oauth2-on-behalf-of-flow/protocols-oauth-on-behalf-of-flow.png).

**Comment by daniellee** [comment]
From @jhunthrop:

To do this properly, we would need to have a much stronger notion of a "user" context in Grafana that we could bind configuration to. Even then, it still doesn't solve the UX issues that @kminehart  raised, which I think is the hardest problem with this. What happens when a user logs into a dashboard and needs to re-authenticate with N number of datasources? Do they go through N different OAuth implicit flows? If we use client credential flow like most integration systems, we end up with a painful user experience getting started with Grafana.

This is a tough problem and will require a lot of thought. 

Looking at other data and orchestration integration products, they commonly take a localized user "authorized app" approach. That makes complete sense to me, but will make this project non-trivial.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
